src/content.js

```diff
@@ -433,3 +435,72 @@ export function nukeTag (host, range, tagName) {
     if (elem.nodeName.toUpperCase() === tagName.toUpperCase()) unwrap(elem)
   })
 }
+
+function createNodeIterator (root, filter) {
+  let currentNode = root
+  let previousNode = null
+
+  function nextNode () {
+    if (!currentNode) {
+      return null
+    }
+
+    if (currentNode.firstChild && previousNode !== currentNode.firstChild) {
+      previousNode = currentNode
+      currentNode = currentNode.firstChild
+    } else if (currentNode.nextSibling) {
+      previousNode = currentNode
+      currentNode = currentNode.nextSibling
+    } else {
+      let parent = currentNode.parentNode
+      while (parent && parent !== root) {
+        if (parent.nextSibling) {
+          previousNode = currentNode = parent.nextSibling
+          break
+        }
+        parent = parent.parentNode
+      }
+      if (!parent || parent === root) {
+        previousNode = currentNode = null
+      }
+    }
+
+    return currentNode
+  }
+
+  return {
+    next: nextNode
+  }
+}
+
+function isNodeFullyContained (node, range) {
+  const nodeRange = document.createRange()
+  nodeRange.selectNodeContents(node)
+  return range.compareBoundaryPoints(Range.START_TO_START, nodeRange) <= 0 &&
+         range.compareBoundaryPoints(Range.END_TO_END, nodeRange) >= 0
+}
+
+function canSurroundContents (range) {
```

Comment:
This does not work properly

Reply:
Can you describe this in a bit more detail so I could test this. Or was it unrelated to this PR anyway (still happy to test if you have use cases in mind)?

Reply:
Ah, I saw that or a very similar bug too!

I'll give it a shot and try to fix it in this PR.

Reply:
It took me a while to understand what the current code in range-save-restore tried to do... I left it mostly in place, added a comment and just fixed the bug 😎